servicetalk-concurrent-api/src/main/java/io/servicetalk/concurrent/api/PublisherFlatMapSingle.java

```diff
@@ -222,23 +223,31 @@ private void enqueueAndDrain(Object item) {
             assert s != null;
 
             if (!pending.offer(item)) {
-                QueueFullException exception = new QueueFullException("pending");
-                if (item instanceof TerminalNotification) {
-                    LOGGER.error("Queue should be unbounded, but an offer failed!", exception);
-                    throw exception;
-                } else {
-                    onError0(exception, true, true);
-                }
+                enqueueAndDrainFail(item);
             }
             drainPending(s);
         }
 
         private void drainPending(Subscription subscription) {
-            long drainedCount = drainSingleConsumerQueue(pending, this::sendToTarget, emittingUpdater, this);
-            if (drainedCount != 0) {
+            long drainCount = 0;
+            for (;;) {
+                if (!acquirePendingLock(emittingUpdater, this)) {
+                    break;
+                }
+                Object t;
+                while ((t = pending.poll()) != null) {
+                    ++drainCount;
+                    sendToTarget(t);
+                }
+                if (releasePendingLock(emittingUpdater, this)) {
```

Comment:
shouldn't we release in a finally block ?

Reply:
good point ... let me make this more robust.